internal/consensus/state.go

```diff
@@ -1377,27 +1389,37 @@ func (cs *State) defaultDoPrevote(height int64, round int32) {
 		return
 	}
 
+	pbtsEnabled := cs.isPBTSEnabled(height)
+	if pbtsEnabled && !cs.Proposal.Timestamp.Equal(cs.ProposalBlock.Header.Time) {
+		logger.Debug("prevote step: proposal timestamp does not match block time; prevoting nil")
+		cs.signAddVote(types.PrevoteType, nil, types.PartSetHeader{}, nil)
+		return
+	}
+
 	// Timestamp validation using Proposed-Based TimeStamp (PBTS) algorithm.
 	// See: https://github.com/cometbft/cometbft/blob/main/spec/consensus/proposer-based-timestamp/README.md
-	if cs.isPBTSEnabled(height) {
-		if !cs.Proposal.Timestamp.Equal(cs.ProposalBlock.Header.Time) {
-			logger.Debug("prevote step: proposal timestamp not equal; prevoting nil")
-			cs.signAddVote(types.PrevoteType, nil, types.PartSetHeader{}, nil)
-			return
-		}
-
-		if cs.Proposal.POLRound == -1 && cs.LockedRound == -1 && !cs.proposalIsTimely() {
-			logger.Debug("prevote step: Proposal is not timely; prevoting nil",
-				"proposed",
-				cmttime.Canonical(cs.Proposal.Timestamp).Format(time.RFC3339Nano),
-				"received",
-				cmttime.Canonical(cs.ProposalReceiveTime).Format(time.RFC3339Nano),
-				"msg_delay",
-				cs.state.ConsensusParams.Synchrony.MessageDelay,
-				"precision",
-				cs.state.ConsensusParams.Synchrony.Precision)
+	if pbtsEnabled && cs.Proposal.POLRound == -1 {
```

Comment:
What happened with `cs.LockedRound == -1`?

Reply:
I deliberately removed it. As it is not part of the PBTS algorithm, and no-one proved me (e.g., via test units) that this condition changes the operation in any way.

Reply:
I dug around a bit:
* `cs.LockedRound == -1` is present on [tip of `v0.36.x`](https://github.com/cometbft/cometbft/blob/v0.36.x/internal/consensus/state.go#L1535)
* By blaming that version, we get it was committed as part of the big branch-squasing PR (tendermint/tendermint#7605) which we forward-ported first. So this confirms our suspicions
* Finally, out of curiosity, I went to branch `wb/pbts-rebase-master` (which looks like the branch tendermint/tendermint#7605 was squashing) and it leads to tendermint/tendermint#7415. We might want to ask @ancazamfir to check we're not missing anything here when she's back.